elastic/kibana
# [Authz] Operator privileges

## Summary

This PR adds support for explicit indication whether endpoint is restricted to operator only users.

### Context
1. If user has [all operator privileges](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/DefaultOperatorOnlyRegistry.java#L35-#L53) granted, but is not listed as operator in `operator_users.yml`, ES would throw an unauthorized error.
2. If user is listed as operator in `operator_users.yml`, but doesn't have necessary privileges granted, ES would throw an unauthorized error.
3. It’s not possible to determine if a user is operator via any ES API, i.e. `_has_privileges`.
4. If operator privileges are disabled we skip the the check for it, that's why we require to explicitly specify additional privileges to ensure that the route is protected even when operator privileges are disabled.

### Checklist


__Relates: https://github.com/elastic/kibana/issues/196271__

### How to test

1. Add your user to the operators list https://github.com/elastic/kibana/blob/1bd81449242a1ab57e82c211808753e82f25c92c/packages/kbn-es/src/serverless_resources/operator_users.yml#L4 or use existing user from the list to log in.
2. Run ES and Kibana serverless
3. Change any endpoint or create a new one with the following security config
```
      security: {
        authz: {
          requiredPrivileges: [ReservedPrivilegesSet.operator],
        },
      },
```
4. Check with enabled and disabled operator privileges (set `xpack.security.operator_privileges.enabled`)

## Release Note
Added support for explicit indication whether endpoint is restricted to operator only users at the route definition level.
